### PR TITLE
Add infrastructure for building coredistools

### PR DESCRIPTION
**Pull request by echesakovMSFT**

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib PTAL

For some reason macOS Azure DevOps hosted pool is super-slow today, so I don't expect the macOS job to succeed. (It was working fine yesterday and the day before).

I will open a meta-issue to track down the remaining work, but this PR should a major un-blocker - the binaries are built for all target platforms and all the host platforms .NET Core supports.

File: build-coredistools.cmd [hunk 1]
```diff
@@ -0,0 +1,66 @@
+@echo off
+setlocal
```
Comment (BruceForstall) [inline build-coredistools.cmd:2]: It's useful to use: `setlocal EnableDelayedExpansion EnableExtensions`

File: build-coredistools.cmd [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
```
Comment (BruceForstall) [inline build-coredistools.cmd:3]: I like to see a usage message at the top, which can double as the documentation for what the script does. E.g.,:

```
@echo off
setlocal

goto start
:usage
echo Usage: build-coredistools.cmd ^<TargetOSArchitecture^>
echo where ^<TargetOSArchitecture^> is one of: windows-arm, windows-arm64, windows-x64, windows-x86.
goto :eof

:start
set RootDirectory=%~dp0
set SourcesDirectory=%RootDirectory%src
set BinariesDirectory=%RootDirectory%obj
set TargetOSArchitecture=%1

if /i "%TargetOSArchitecture%" == "windows-arm" (
    set GeneratorPlatform=ARM
) else if /i "%TargetOSArchitecture%" == "windows-arm64" (
    set GeneratorPlatform=ARM64
) else if /i "%TargetOSArchitecture%" == "windows-x64" (
    set GeneratorPlatform=x64
) else if /i "%TargetOSArchitecture%" == "windows-x86" (
    set GeneratorPlatform=Win32
) else (
    echo "Unknown target OS and architecture: %TargetOSArchitecture%"
    goto usage ///********
)
```

File: build-coredistools.cmd [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+set RootDirectory=%~dp0
+set SourcesDirectory=%RootDirectory%src
+set BinariesDirectory=%RootDirectory%obj
+set TargetOSArchitecture=%1
+
+if /i "%TargetOSArchitecture%" == "windows-arm" (
```
Comment (BruceForstall) [inline build-coredistools.cmd:9]: Could this default to using `PROCESSOR_ARCHITECTURE`?

File: build-coredistools.cmd [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+cmake.exe ^
+    -G "Visual Studio 16 2019" ^
+    -A %GeneratorPlatform% ^
+    -DCMAKE_INSTALL_PREFIX="%RootDirectory%\" ^
+    -DLLVM_EXTERNAL_PROJECTS=coredistools ^
+    -DLLVM_EXTERNAL_COREDISTOOLS_SOURCE_DIR="%SourcesDirectory%\coredistools" ^
+    -DLLVM_TABLEGEN="%LLVMTableGen%" ^
+    -DLLVM_TARGETS_TO_BUILD=AArch64;ARM;X86 ^
+    -DLLVM_TOOL_COREDISTOOLS_BUILD=ON ^
+    "%SourcesDirectory%\llvm-project\llvm"
```
Comment (BruceForstall) [inline build-coredistools.cmd:48]: You should check the ERRORLEVEL return for all 3 cmake invocations, not just one.

File: build-coredistools.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        CrossCompiling=1
+        TargetTriple=aarch64-linux-gnu
+        ;;
+
+    linux-x64|macos-x64)
+        CrossCompiling=0
+        ;;
+
+    *)
+        echo "Unknown target OS and architecture: $TargetOSArchitecture"
```
Comment (BruceForstall) [inline build-coredistools.sh:25]: A similar comment applies here about it being useful to have a "help" / usage comment and, if necessary and separate, a "what does this script do" header comment.

File: build-tblgen.cmd [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+set BinariesDirectory=%RootDirectory%obj
+set StagingDirectory=%RootDirectory%bin
+
+if not exist "%BinariesDirectory%" (
+    mkdir "%BinariesDirectory%"
+)
+
+pushd "%BinariesDirectory%"
+
+cmake.exe ^
```
Comment (BruceForstall) [inline build-tblgen.cmd:15]: check all the cmake exit codes?

File: build-tblgen.cmd [hunk 7]
```diff
@@ -0,0 +1,45 @@
+@echo off
+setlocal
+
```
Comment (BruceForstall) [inline build-tblgen.cmd:3]: Add a "what does this script do?" header comment?

You could also have:
```
if not "%1"=="" echo Error: no arguments expected&goto :eof
```

**Review by BruceForstall — APPROVED** [review]
LGTM. I only had minor suggestions for improving the scripts (which I realized after writing the comments would probably only be invoked by the YML, not manually, but still...)

**Comment by echesakovMSFT** [comment]
Thanks @BruceForstall for code review! I will have a follow-up PR when I address some of the issues with Arm32 I am working on and all of your suggestions here.

However, I would like to have a working build pipeline at the moment when I open that PR, so I am merging these changes.

**Pull request merged**